File: src/simulation/culture/group/TerritoryCenter.java
package simulation.culture.group;

import extra.SpaceProbabilityFuncs;
import simulation.Controller;
import simulation.Event;
import simulation.space.Territory;
import simulation.space.tile.Tile;
import simulation.space.tile.TileDistanceKt;

import java.util.*;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static shmp.random.RandomProbabilitiesKt.*;

public class TerritoryCenter {
    private Territory territory = new Territory();
    private GroupTileTag tileTag;
    private Function<Tile, Integer> tilePotentialMapper = t ->
            t.getNeighbours(tile1 -> tile1.getTagPool().contains(tileTag)).size()
                    + 3 * t.getResourcePack().getAmount(r ->
                    tileTag.getGroup().getCultureCenter().getAspectCenter().getAspectPool().getResourceRequirements()
                            .contains(r)
            )
                    + tileTag.getGroup().getRelationCenter().evaluateTile(t);
    private double spreadAbility;

    TerritoryCenter(Group group, double spreadAbility, Tile tile) {
        this.tileTag = new GroupTileTag(group);
        this.spreadAbility = spreadAbility;
        claimTile(tile);
    }

    public Territory getTerritory() {
        return territory;
    }

    public double getSpreadAbility() {
        return spreadAbility;
    }

    public Function<Tile, Integer> getTilePotentialMapper() {
        return tilePotentialMapper;
    }

    public Set<Group> getAllNearGroups(Group exception) {
        Set<Group> groups = territory.getOuterBrink().stream()
                .flatMap(t -> t.getTagPool().getByType(tileTag.getType()).stream())
                .map(t -> ((GroupTileTag) t).getGroup())
                .collect(Collectors.toSet());
        groups.remove(exception);
        return groups;
    }

    boolean migrate() {
        Tile newCenter = getMigrationTile();
        if (newCenter == null) {
            return false;
        }
        getTerritory().setCenter(newCenter);
        claimTile(newCenter);//TODO move claim and leave here
        leaveTiles(getTerritory().getTiles(tile -> !isTileReachable(tile)));
        return true;
    }

    private Tile getMigrationTile() {
        return getMigrationTiles()
                .stream()
                .max(Comparator.comparingInt(tile -> tilePotentialMapper.apply(tile)))
                .orElse(null);
    }

    private Collection<Tile> getMigrationTiles() {
        Set<Tile> tiles = new HashSet<>();
        Set<Tile> queue = new HashSet<>();
        queue.add(territory.getCenter());
        while (true) {
            tiles.addAll(queue);
            List<Tile> currentTiles = queue.stream()
                    .flatMap(t -> t.getNeighbours().stream())
                    .distinct()
                    .filter(t -> !tiles.contains(t))
                    .filter(this::isTileReachable)
                    .filter(this::canTraverse)
                    .collect(Collectors.toList());
            if (currentTiles.isEmpty()) {
                break;
            }
            queue.clear();
            queue.addAll(currentTiles);
        }
        return tiles.stream()
                .filter(this::canSettleAndNoGroup)
                .collect(Collectors.toList());
    }

    private boolean canTraverse(Tile tile) {
        if (tile.getType() == Tile.Type.Water) {
            return false;
        } else if (tile.getType() == Tile.Type.Mountain) {
            if (!tileTag.getGroup().getCultureCenter().getAspectCenter().getAspectPool().contains("MountainLiving")) {
                return false;
            }
        }
        return true;
    }

    public Tile getDisbandTile() {
        return SpaceProbabilityFuncs.randomTile(getTerritory());
    }

    boolean expand() {
        if (!testProbability(spreadAbility, Controller.session.random)) {
            return false;
        }
        claimTile(territory.getMostUsefulTileOnOuterBrink(
                t -> canSettleAndNoGroup(t) && isTileReachable(t),
                tilePotentialMapper
        ));
        return true;
    }

    void shrink() {
        if (territory.size() <= 1) {
            return;
        }
        leaveTile(territory.getMostUselessTile(tilePotentialMapper));
    }

    private boolean isTileReachable(Tile tile) {
        return TileDistanceKt.getDistance(tile, getTerritory().getCenter()) < 4;
    }

    void claimTile(Tile tile) {
        if (tile == null) {
            return;
        }
        if (!tile.getTagPool().contains(tileTag) && !tile.getTagPool().getByType(tileTag.getType()).isEmpty()) {
            throw new RuntimeException();
        }
        tileTag.getGroup().getParentGroup().claimTile(tile);
        tile.getTagPool().add(tileTag);
        getTerritory().add(tile);
        tileTag.getGroup().addEvent(new Event(
                Event.Type.TileAcquisition,
                "Group " + tileTag.getGroup().name + " claimed tile " + tile.getX() + " " + tile.getY(),
                "group", this, "tile", tile
        ));
    }

    void die() {
        for (Tile tile : getTerritory().getTiles()) {
            tile.getTagPool().remove(tileTag);
        }
    }

    private void leaveTile(Tile tile) {
        if (tile == null) {
            return;
        }
        tile.getTagPool().remove(tileTag);
        getTerritory().removeTile(tile);
        tileTag.getGroup().getParentGroup().removeTile(tile);
    }

    private void leaveTiles(Collection<Tile> tiles) {
        tiles.forEach(this::leaveTile);
    }

    public boolean canSettle(Tile tile) {
        return tile.getType() != Tile.Type.Water && tile.getType() != Tile.Type.Mountain
                || (tile.getType() == Tile.Type.Mountain
                && !tileTag.getGroup().getCultureCenter().getAspectCenter().getAspectPool().filter(a ->
                a.getTags().stream()
                        .anyMatch(aspectTag -> aspectTag.name.equals("mountainLiving")))//TODO set of accessible tiles
                .isEmpty());
    }

    boolean canSettleAndNoGroup(Tile tile) {
        return canSettle(tile, t -> t.getTagPool().getByType(tileTag.getType()).isEmpty());
    }

    public boolean canSettle(Tile tile, Predicate<Tile> additionalCondition) {
        return canSettle(tile) && additionalCondition.test(tile);
    }
}

File: src/simulation/culture/group/TerritoryCenter.kt
package simulation.culture.group

import extra.SpaceProbabilityFuncs
import shmp.random.testProbability
import simulation.Controller
import simulation.Event
import simulation.space.Territory
import simulation.space.tile.Tile
import simulation.space.tile.getDistance
import java.util.function.Consumer
import java.util.function.Predicate

class TerritoryCenter(group: Group, val spreadAbility: Double, tile: Tile) {
    val territory = Territory()

    private val tileTag: GroupTileTag = GroupTileTag(group)

    private var _oldCenter: Tile? = null
    private var _oldReach: Collection<Tile> = listOf()
    private var _oldTileTypes: Collection<Tile.Type> = listOf()

    fun tilePotentialMapper(tile: Tile): Int {
        val convexPart = tile.getNeighbours { it.tagPool.contains(tileTag) }.size
        val neededResourcePart = 3 * tile.resourcePack.getAmount {
            tileTag.group.cultureCenter.aspectCenter.aspectPool.getResourceRequirements().contains(it)
        }
        val relationPart = tileTag.group.relationCenter.evaluateTile(tile)
        return convexPart + neededResourcePart + relationPart
    }

    init {
        claimTile(tile)
    }

    val accessibleTerritory: Territory
        get() = Territory(reachableTiles)

    fun getAllNearGroups(exception: Group) = territory.outerBrink
            .mapNotNull { getResidingGroup(it) }
            .filter { it != exception }
            .toSet()

    val disbandTile: Tile
        get() = SpaceProbabilityFuncs.randomTile(territory)

    fun migrate(): Boolean {
        val newCenter = migrationTile ?: return false
        territory.center = newCenter
        claimTile(newCenter) //TODO move claim and leave here
        leaveTiles(territory.getTiles { !isTileReachable(it) })
        return true
    }

    private val migrationTile: Tile?
        get() = reachableTiles.maxBy { tilePotentialMapper(it) }

    private val reachableTiles: Collection<Tile>
        get() {
            val tileTypes = getAccessibleTileTypes()
            if (_oldCenter != territory.center || tileTypes != _oldTileTypes) {
                _oldCenter = territory.center
                _oldReach = getReachableTiles(territory.center)
                _oldTileTypes = tileTypes
            }
            return _oldReach.filter { canSettleAndNoGroupExcept(it) }
        }

    private fun getAccessibleTileTypes(): List<Tile.Type> {
        return if (tileTag.group.cherishedResources.resources.any { it.simpleName == "Boat"})
            listOf(Tile.Type.Water)
        else listOf()
    }

    private fun getReachableTiles(tile: Tile): Collection<Tile> {
        val tiles = mutableSetOf<Tile>()
        val queue = mutableSetOf<Pair<Tile, Int>>()
        queue.add(tile to 0)
        var currentLayer = 1
        while (true) {
            tiles.addAll(queue.map { it.first })
            val currentTiles = queue
                    .filter { isTileReachableInTraverse(it) }//TODO another sequence
                    .flatMap { it.first.neighbours }
                    .asSequence()
                    .distinct()
                    .filter { !tiles.contains(it) }
                    .filter { canTraverse(it) }
                    .map { it to currentLayer }
                    .toList()
            if (currentTiles.isEmpty()) {
                break
            }
            queue.clear()
            queue.addAll(currentTiles)
            currentLayer++
        }
        return tiles
    }

    private fun isTileReachableInTraverse(pair: Pair<Tile, Int>) = when (pair.first.type) {
        Tile.Type.Water -> if (_oldTileTypes.contains(Tile.Type.Water))
            pair.second <= 30
        else false
        else -> pair.second <= 5
    }

    private fun canTraverse(tile: Tile): Boolean {
        if (tile.type == Tile.Type.Water) {
            return _oldTileTypes.contains(Tile.Type.Water)
        } else if (tile.type == Tile.Type.Mountain) {
            if (!tileTag.group.cultureCenter.aspectCenter.aspectPool.contains("MountainLiving")) {
                return false
            }
        }
        return true
    }

    fun expand(): Boolean {
        if (!testProbability(spreadAbility, Controller.session.random)) {
            return false
        }
        claimTile(territory.getMostUsefulTileOnOuterBrink(
                { canSettleAndNoGroup(it) && isTileReachable(it) },
                this::tilePotentialMapper
        ))
        return true
    }

    fun shrink() {
        if (territory.size() <= 1) {
            return
        }
        leaveTile(territory.getMostUselessTile(this::tilePotentialMapper))
    }

    private fun isTileReachable(tile: Tile) = getDistance(tile, territory.center) < 4

    fun claimTile(tile: Tile?) {
        if (tile == null) {
            return
        }
        if (!tile.tagPool.contains(tileTag) && hasResidingGroup(tile)) {
            throw RuntimeException()
        }
        tileTag.group.parentGroup.claimTile(tile)
        tile.tagPool.add(tileTag)
        territory.add(tile)
        tileTag.group.addEvent(Event(
                Event.Type.TileAcquisition,
                "Group " + tileTag.group.name + " claimed tile " + tile.x + " " + tile.y,
                "group", this, "tile", tile
        ))
    }

    fun die() {
        for (tile in territory.tiles) {
            tile.tagPool.remove(tileTag)
        }
    }

    private fun leaveTile(tile: Tile?) {
        if (tile == null) {
            return
        }
        tile.tagPool.remove(tileTag)
        territory.removeTile(tile)
        tileTag.group.parentGroup.removeTile(tile)
    }

    private fun leaveTiles(tiles: Collection<Tile>) {
        tiles.forEach(Consumer { tile: Tile? -> leaveTile(tile) })
    }

    fun canSettle(tile: Tile) = (tile.type != Tile.Type.Water && tile.type != Tile.Type.Mountain
            || (tile.type == Tile.Type.Mountain//TODO set of accessible tile types
            && tileTag.group.cultureCenter.aspectCenter.aspectPool.contains("mountainLiving")))

    fun canSettleAndNoGroup(tile: Tile) =
            canSettle(tile, Predicate { hasNoResidingGroup(it) })


    fun canSettleAndNoGroupExcept(tile: Tile) =
            canSettle(tile, Predicate { hasNoResidingGroupExcept(it, tileTag.group) })

    fun canSettle(tile: Tile, additionalCondition: Predicate<Tile>) =
            canSettle(tile) && additionalCondition.test(tile)
}